api/task_processor/models.py

```diff
@@ -0,0 +1,89 @@
+import json
+import typing
+import uuid
+from datetime import datetime
+
+from django.db import models
+from django.utils import timezone
+
+from task_processor.exceptions import TaskProcessingError
+from task_processor.tasks import registered_tasks
+
+
+class Task(models.Model):
+    uuid = models.UUIDField(unique=True, default=uuid.uuid4)
+    created_at = models.DateTimeField(auto_now_add=True)
+    scheduled_for = models.DateTimeField(blank=True, null=True, default=timezone.now)
+
+    task_identifier = models.CharField(max_length=200)
+    serialized_args = models.TextField(blank=True, null=True)
+    serialized_kwargs = models.TextField(blank=True, null=True)
+
+    # denormalise failures so that we can use select_for_update
+    num_failures = models.IntegerField(blank=True, null=True, default=0)
```

Comment:
I think perhaps this field should have an index, so that the task processor query can efficiently find tasks to process and ignore the (even growing) list of processed tasks.

Reply:
Yep, good shout. 

Since the filter query looks something like the following, I've actually added a combined index on both `num_failures` and `scheduled_for`. Doing some postgres explains seems to show it's the most efficient option. 

```python
.filter(num_failures__lt=3, scheduled_for__lte=timezone.now())
```